### DISCUSSION TITLE
Support relative max data points

### DISCUSSION THREAD

**Discussion by breezewish** [grafana/grafana#50134] [Dashboards, Panels & Visualizations]
Update: A demo of this proposal has been implemented in https://github.com/grafana/grafana/pull/50214

<img width="476" alt="image" src="https://user-images.githubusercontent.com/1916485/171783453-300d5724-2988-4243-bc27-3dc42efe37fa.png">

The max data points options now only have two options:

- An absolute number of data points
- Auto, equals to the panel visible width

It would be better to additionally support percentage notation, e.g. specifying `30%` in this option. This enables user to control **how "dense" the chart should be displayed** dynamically. Specifying `100%` would be just identical to specifying `auto`.

Note that a `50%` max data point is not identical to set resolution to `1/2`:

Consider that we have a metric that usually changes sharply. When the time range is large, e.g. 1 day, this chart is hard to read due to the density of the data points:

<img width="639" alt="image" src="https://user-images.githubusercontent.com/1916485/171783888-18f71c4b-0528-4030-9179-913f7612cb70.png">

User could possibly improve it by setting to a lower resolution, e.g. `1/10`:

<img width="637" alt="image" src="https://user-images.githubusercontent.com/1916485/171784217-97dbc163-b0a8-446d-84e4-0fceb35eb267.png">

However, this resolution change causes new problems when user want to view details, e.g. by selecting a 30 minute time range:

<img width="637" alt="image" src="https://user-images.githubusercontent.com/1916485/171784526-03043c12-3f32-432e-918d-dfa88f911e8f.png">

User cannot view the exact detail, due to the low resolution, even when the panel is large and more data points can be displayed in the screen.

An ideal charts looks like this, by displaying more details to users:

<img width="641" alt="image" src="https://user-images.githubusercontent.com/1916485/171784710-f9378bb9-c64c-439a-859c-5d1a2c881f2e.png">

This can only be achieved by setting the resolution back to `1/1`. This UX is not good for users, as users must additionally choose a good enough "resolution" when changing time ranges, in order to view these charts correctly and nicely.

This problem can be resolved by allowing to set a relative "max data points". For example, the user could use this option:

- Max data points = 10%
- Resolution = 1/1

The produced chart now works great in both cases: when the time range is large (result in 90% less data points) and the time range is small (result in all data points being displayed).

**Comment by belm0** [comment]
> When the time range is large, e.g. 1 day, this chart is hard to read due to the density of the data points:

I think the main use case for controlling the data point count is to reduce load on the client, server, backend storage, and network.

If you need series to be smoothed for visual readability, use `avg_over_time()`, etc.

**Reply by breezewish** [reply]
> > When the time range is large, e.g. 1 day, this chart is hard to read due to the density of the data points:
> 
> I think the main use case for controlling the data point count is to reduce load on the client, server, backend storage, and network.
> 
> If you need series to be smoothed for visual readability, use `avg_over_time()`, etc.

Yeah, the 10% max data points should be usually used with functions like `avg_over_time`.

This shows how 10% max data points differs from the resolution:

**Large Time Range**, Series are smoothed:

<img width="1431" alt="image" src="https://user-images.githubusercontent.com/1916485/172034729-522585a5-73bf-4f82-a1bc-3dfef9131b95.png">

**Small Time Range**, Details like jitters are still displayed:

<img width="1431" alt="image" src="https://user-images.githubusercontent.com/1916485/172034735-82cf0bba-4231-4928-a0e6-608bcfe759a3.png">

The relative max data points produce great visualizations, **in both cases**, when user selected a large time range that contains many data points, or selected a small time range. The details would not be sacrificed.

This cannot be achieved by setting resolution or specifying a fixed interval for functions like `avg_over_time`.

What's more, **it also works smoothly when the panel size is changed**!

For small panels, series are smoothed, and for large panels that have enough space to display, details are revealed:

<img width="1434" alt="image" src="https://user-images.githubusercontent.com/1916485/172034781-4670f576-f2e3-441d-929a-4afb88b921e7.png">

**Reply by belm0** [reply]
> This cannot be achieved by setting resolution or specifying a fixed interval for functions like `avg_over_time`.

there is a variable `$__interval` that may be suitable for functions like `avg_over_time()`

https://grafana.com/docs/grafana/latest/variables/variable-types/global-variables/#__interval

`Approximate Calculation: (to - from) / resolution`

**Reply by breezewish** [reply]
I'm already using `$__rate_interval` for all charts in my posted image. The reason why `Resolution = 1/1` looks not so nice is because Grafana attempts to display too many data points, i.e. 1 data point per pixel. This is not suitable for charts that varies drastically. By specifying `max data points = 10%`, Grafana will display 1 data point **per 10 pixels** -- smoothed.

**Comment by srclosson** [comment]
Hmmmm..... some thoughts.

When asking for a smaller resolution, what does this actually mean? 1/10th of all the data points is fine. Maybe this means we send only data poing 0, 10, 20, 30, 40 etc... but what this seems to boil down to is an aggregation then. What are the insights being looked for? The max, the min, the average? Query languages should be solving this problem. Not the panel configuration.

**Reply by belm0** [reply]
Queries should not be written expecting that they have 1:1 data points at the client.

The query decides whether it will summarize between display points or not, and how (max, avg, etc.).  Grafana gives guidance on what the interval should be, see `$__interval`.

https://grafana.com/docs/grafana/latest/variables/variable-types/global-variables/#__interval

**Reply by breezewish** [reply]
I agree with @belm0. The Query should always expecting a non 1:1 data points, even when this feature is not yet supported. An auto "max data points" already pose Prometheus step to be larger than min-step when selected time range is large. In order to display correctly **for any selected time range**, user must use $__interval, or $__rate_interval with some functions like `rate()` or `avg_over_time()`.

**Comment by vinnitu** [comment]
Hello. Can I join to your talk?

I liked your phase @breezewish - "...in order to view these charts correctly and nicely"

Imho the best way to do it use dataPoints value to proper calculation interval_ms value
![image](https://user-images.githubusercontent.com/432168/175499673-9f1635cc-5ee3-4ac6-acac-7bd41d736974.png)

Which in this case will be - equal 30s / 803 px.
It means we need make correct aggregated query to database with this interval (~37.36 ms per pixel).
Because it is not integer value - we need pass fraction (both params) - time range (30s) and chart width (803px)

![image](https://user-images.githubusercontent.com/432168/175499619-0327b5c3-0422-4984-bf8e-c8a4f22d1a3f.png)
Value of maxDataPoints we can use only to limit query response (for example only first 100 points).
But in general maxDataPoints is useless.

And it is relative to my [question](https://github.com/grafana/grafana/discussions/51363)

**Reply by breezewish** [reply]
Indeed! Using the width of the chart (instead of the width of the panel) is more reasonable.

**Reply by torkelo** [reply]
Grafana query runner does not know the exact width of the chart (available for lines). As it depends on number of y axis, and y axis width (which can vary depending on data). 

But we can subtract the padding at least.

**Comment by vinnitu** [comment]
because interval_ms calculated by grafana is not equal (timeRange / dataPoints) we can see blinking effect like this
![Peek 2022-06-24 12-19](https://user-images.githubusercontent.com/432168/175505715-203e9479-626b-465f-aeea-9987bb6e0019.gif)

**Comment by torkelo** [comment]
@breezewish thank you for your really well-formulated suggestion, I think it makes some sense. 

But I think resolution and percentage would be the same if they was applied to max data points. I think the reason your seeing wrong resolution (the prometheus setting) behavior is that it acts on interval not max data points. It ranges from 1 to 10, and just multiples the interval (step) always. If it instead was 1/1  to 1/10 multiplier on the max data points property it would work as you suggest.

**Comment by torkelo** [comment]
https://github.com/grafana/grafana/pull/51687

**Comment by natellium** [comment]
Hey there, as we're moving back to GH issue I've converted this one https://github.com/grafana/grafana/issues/78528
